worlds/lingo/regions.py

```diff
@@ -0,0 +1,78 @@
+    new_region = Region(room.name, player, multiworld)
+
+    if room.name in player_logic.LOCATIONS_BY_ROOM.keys():
+        for location in player_logic.LOCATIONS_BY_ROOM[room.name]:
+            new_loc = LingoLocation(player, location.name, location.code, new_region)
+            new_region.locations.append(new_loc)
+
+    multiworld.regions += [
+        new_region
+    ]
+
+
+def connect(target: Room, entrance: RoomEntrance, multiworld: MultiWorld, player: int, player_logic: LingoPlayerLogic):
+    target_region = multiworld.get_region(target.name, player)
+    source_region = multiworld.get_region(entrance.room, player)
+    connection = Entrance(player, f"{entrance.room} to {target.name}", source_region)
+    connection.access_rule = lambda state: state.lingo_can_use_entrance(
+        target.name, entrance.door, player, player_logic)
+
+    source_region.exits.append(connection)
+    connection.connect(target_region)
+
+
+def handle_pilgrim_room(multiworld: MultiWorld, player: int, player_logic: LingoPlayerLogic):
+    target_region = multiworld.get_region("Pilgrim Antechamber", player)
+    source_region = multiworld.get_region("Outside The Agreeable", player)
+    connection = Entrance(player, f"Pilgrimage", source_region)
+    connection.access_rule = lambda state: state.lingo_can_use_pilgrimage(
+        player, player_logic)
+
+    source_region.exits.append(connection)
+    connection.connect(target_region)
+
+
+def connect_painting(warp_enter: str, warp_exit: str, multiworld: MultiWorld, player: int,
+                     static_logic: StaticLingoLogic, player_logic: LingoPlayerLogic):
+    source_painting = static_logic.PAINTINGS[warp_enter]
+    target_painting = static_logic.PAINTINGS[warp_exit]
+
+    target_region = multiworld.get_region(target_painting.room, player)
+    source_region = multiworld.get_region(source_painting.room, player)
+    connection = Entrance(player, f"{source_painting.room} to {target_painting.room} (Painting)", source_region)
+    connection.access_rule = lambda state: state.lingo_can_use_entrance(
+        target_painting.room, source_painting.required_door, player, player_logic)
+
+    source_region.exits.append(connection)
+    connection.connect(target_region)
+
+
+def create_regions(multiworld: MultiWorld, player: int, static_logic: StaticLingoLogic, player_logic: LingoPlayerLogic):
+    multiworld.regions += [
+        Region("Menu", player, multiworld)
+    ]
+
+    for room in static_logic.ALL_ROOMS:
+        create_region(room, multiworld, player, player_logic)
+
+    for room in static_logic.ALL_ROOMS:
+        for entrance in room.entrances:
+            if entrance.painting and get_option_value(multiworld, player, "shuffle_paintings"):
+                # Don't use the vanilla painting connections if we are shuffling paintings.
+                continue
+
+            connect(room, entrance, multiworld, player, player_logic)
```

Comment:
This doesn't work here, because the lambda holds on to the references to the loop variables, which get overwritten after each iteration. I've applied this suggestion within connect and the other functions, though, to make them simpler.